onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc

```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
+    const auto input0_info = graph.GetValueInfo(inputs[0]);
+    auto input0_rank = input0_info->ShapeRank();
+    if (!input0_rank.has_value() || !NormalizeAndValidateAxis(axis, *input0_rank)) {
+      return std::nullopt;  // Unable to normalize the DQ's axis.
+    }
+
+    int64_t block_size = q_or_dq_node.GetAttributeIntDefault("block_size", 0);
+    if (block_size != 0) {
+      // A block size indicates blocked quantization.
+      quant_info.mode = QuantizationMode::kBlocked;
+      quant_info.norm_axis = axis;
+    } else {
+      // A block size of 0 indicates per-axis quantization.
+      quant_info.mode = QuantizationMode::kPerAxis;
+      quant_info.norm_axis = axis;
+    }
+  }
+
+  return quant_info;
+}
+
+/// <summary>
+/// Represents a DequantizeLinear node that TransposeInputImpl() or UnsqueezeInput() can look past to transpose or
+/// unsqueeze its input.
+/// </summary>
+class DQToLookPast {
+ public:
+  DQToLookPast() = default;
+  DQToLookPast(std::unique_ptr<api::NodeRef>&& dq_node, QuantizationInfo quant_info)
+      : dq_node_(std::move(dq_node)), quant_info_(quant_info) {
+  }
+
+  DQToLookPast(DQToLookPast&& other) = default;
+  DQToLookPast& operator=(DQToLookPast&& other) = default;
+
+  inline QuantizationMode GetQuantMode() const {
+    return quant_info_.mode;
+  }
+
+  inline void DisconectInput0() {
+    dq_node_->SetInput(0, "");
+  }
+
+  inline void ReconnectInput0(std::string_view input_name) {
+    dq_node_->SetInput(0, input_name);
+  }
+
+  inline std::string_view GetInput0() {
+    return dq_node_->Inputs()[0];
+  }
+
+  inline std::string_view GetOutput() {
+    return dq_node_->Outputs()[0];
+  }
+
+  /// <summary>
+  /// Sets the DQ's new transposed input[0]. The DQ's axis and output shape are updated.
+  /// </summary>
+  /// <param name="graph">graph</param>
+  /// <param name="new_input">name of new transposed input[0]</param>
+  /// <param name="perm_inv">inverse transpose permutation used to update the DQ's axis</param>
+  void SetTransposedInput(api::GraphRef& graph, std::string_view new_input, gsl::span<const int64_t> perm_inv) {
+    int64_t axis = quant_info_.norm_axis;
+    if (quant_info_.mode == QuantizationMode::kPerAxis) {
+      axis = perm_inv[gsl::narrow_cast<size_t>(axis)];
+    }
+
+    dq_node_->SetInput(0, new_input);
+    dq_node_->SetAttributeInt("axis", axis);
+    auto new_shape = *graph.GetValueInfo(new_input)->Shape();
+    graph.GetValueInfo(dq_node_->Outputs()[0])->SetShape(&new_shape);
+  }
+
+  /// <summary>
+  /// Sets the DQ's new unsqueezed input[0]. The DQ's axis and output shape are updated.
+  /// The provided unsqueeze axes must all be positive (i.e., normalized).
+  /// </summary>
+  /// <param name="graph">graph</param>
+  /// <param name="new_input">name of new unsqueezed input[0]</param>
+  /// <param name="perm_inv">positive unsqueeze axes used to update the DQ's axis</param>
+  void SetUnsqueezedInput(api::GraphRef& graph, std::string_view new_input,
```

Comment:
nit: the impl for this and SetTransposedInput look to be the same so maybe we could split out into a helper func? happy to have both these functions calling the helper for the clarify it adds. will reduce binary size by a tiny bit as well.

Reply:
Sounds good. I split out common logic into a private helper.